internal/runners/push/push.go

```diff
@@ -42,146 +43,189 @@ type primeable interface {
-		if err != nil {
-			return errs.Wrap(err, "Could not get default branch")
-		}
 
 		projectCreated = true
 	}
 
-	var commitID = r.project.CommitUUID()
-	if commitID.String() == "" {
-		var err error
-		commitID, err = model.CommitInitial(model.HostPlatform, lang, langVersion)
+	// Now we get to the actual push logic
+	r.out.Notice(locale.Tl("push_to_project", "Pushing to project [NOTICE]{{.V1}}[/RESET] under [NOTICE]{{.V0}}[/RESET].", targetNamespace.Owner, targetNamespace.Project))
+
+	// Detect the target branch
+	var branch *mono_models.Branch
+	if r.project.BranchName() == "" {
+		// https://www.pivotaltracker.com/story/show/176806415
+		branch, err = model.DefaultBranchForProject(targetPjm)
 		if err != nil {
-			return locale.WrapError(err, "push_project_init_err", "Failed to initialize project {{.V0}}", pjm.Name)
+			return locale.NewInputError("err_no_default_branch")
+		}
+	} else {
+		branch, err = model.BranchForProjectByName(targetPjm, r.project.BranchName())
+		if err != nil {
+			return locale.WrapError(err, "err_fetch_branch", "", r.project.BranchName())
 		}
 	}
 
-	// update the project at the given commit id.
-	err = model.UpdateProjectBranchCommitWithModel(pjm, branch.Label, commitID)
-	if err != nil {
-		return locale.WrapError(err, "push_project_branch_commit_err", "Failed to update new project {{.V0}} to current commitID.", pjm.Name)
+	// Check if branch is already up to date
+	if branch.CommitID != nil && branch.CommitID.String() == commitID.String() {
+		r.out.Notice(locale.T("push_no_changes"))
+		return nil
 	}
 
-	// Remove temporary language entry
-	pjf := r.project.Source()
-	err = pjf.RemoveTemporaryLanguage()
+	// Check whether there is a conflict
+	if branch.CommitID != nil {
+		mergeStrategy, err := model.MergeCommit(*branch.CommitID, commitID)
+		if err != nil {
+			if errors.Is(err, model.ErrMergeCommitInHistory) {
+				r.out.Notice(locale.T("push_no_changes"))
+				return nil
+			}
+			if !errors.Is(err, model.ErrMergeFastForward) {
+				if params.Namespace.IsValid() {
+					return locale.WrapError(err, "err_mergecommit_customtarget", "The targets commit history does not match your local commit history.")
+				}
+				return locale.WrapError(err, "err_mergecommit", "Could not detect if merge is necessary.")
+			}
+		}
+		if mergeStrategy != nil {
+			return errs.AddTips(
+				locale.NewInputError("err_push_outdated"),
+				locale.Tl("err_tip_push_outdated", "Run `[ACTIONABLE]state pull[/RESET]`"))
+		}
+	}
+
+	// Update the project at the given commit id.
+	err = model.UpdateProjectBranchCommitWithModel(targetPjm, branch.Label, commitID)
 	if err != nil {
-		return locale.WrapInputError(err, "push_remove_lang_err", "Failed to remove temporary language field from activestate.yaml.")
+		if errs.Matches(err, &model.ErrUpdateBranchAuth{}) {
+			return locale.WrapInputError(err, "push_project_branch_no_permission", "You do not have permission to push to {{.V0}}.", targetPjm.Name)
```

Comment:
Does `targetPjm.Name` include the organization name? If not we may want to include it in the errors here.